File: 34.find-first-and-last-position-of-element-in-sorted-array.py
#
# @lc app=leetcode id=34 lang=python3
#
# [34] Find First and Last Position of Element in Sorted Array
#
from bisect import bisect_left, bisect_right
from typing import *

# @lc code=start
class Solution:
    # T: O(log(n))
    # S: O(1)
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        n = len(nums)

        # Find left most
        l, r = 0, n - 1
        while l <= r:
            m = l + ((r - l) >> 1)
            if nums[m] >= target:
                r = m - 1
            else:
                l = m + 1
        if l >= n or nums[l] != target:
            return [-1, -1]
        ans_l = l

        # Find right most 
        # (note that setting l = 0 is not necessary if two searches are consecutive)
        l, r = 0, n - 1
        while l <= r:
            m = l + ((r - l) >> 1)
            if nums[m] <= target:
                l = m + 1
            else:
                r = m - 1
        if r < 0 or nums[r] != target:
            return [-1, -1]
        return [ans_l, r]

    """
    # Solution with built-in functions
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if not len(nums):
            return [-1, -1]
        l = bisect_left(nums, target)
        r = bisect_right(nums, target)
        return [l, r - 1] if l < r else [-1, -1]
    """
        
# @lc code=end

File: 34.find-first-and-last-position-of-element-in-sorted-array.java
/*
 * @lc app=leetcode id=34 lang=java
 *
 * [34] Find First and Last Position of Element in Sorted Array
 */

// @lc code=start
class Solution {
    // T: O(log(n))
    // S: O(1)
    public int[] searchRange(int[] nums, int target) {
        int[] result = new int[2];
        result[0] = findFirst(nums, target);
        result[1] = findLast(nums, target);
        return result;
    }
    
    private int findFirst(int[] nums, int target) {
        int idx = -1;
        int l = 0;
        int r = nums.length - 1;
        while (l <= r) {
            int m = l + ((r - l) >>> 1);
            if (nums[m] >= target) {
                r = m - 1;
            } else {
                l = m + 1;
            }
            if (nums[m] == target) {
                idx = m;
            }
        }
        return idx;
    }
    
    private int findLast(int[] nums, int target) {
        int idx = -1;
        int l = 0;
        int r = nums.length - 1;
        while (l <= r) {
            int m = l + ((r - l) >>> 1);
            if (nums[m] <= target) {
                l = m + 1;
            } else {
                r = m - 1;
            }
            if (nums[m] == target) {
                idx = m;
            }
        }
        return idx;
    }
}
// @lc code=end

